dewy/common/collection_embeddings.py

```diff
@@ -0,0 +1,259 @@
+
+        # TODO: Ideally the collection embeddings would be cached, and this
+        # wouldn't need to exist.
+        async with pg_pool.acquire() as conn:
+            result = await conn.fetchrow(
+                """
+                SELECT
+                    document.id,
+                    document.url,
+                    collection.name,
+                    collection.id as collection_id,
+                    collection.text_embedding_model,
+                    collection.text_distance_metric
+                FROM document
+                JOIN collection ON document.collection_id = collection.id
+                WHERE document.id = $1;
+                """,
+                document_id,
+            )
+
+            # TODO: Cache the configured ingestions, and only recreate when needed?
+            configured_ingestion = CollectionEmbeddings(pg_pool, result)
+            return (result["url"], configured_ingestion)
+
+    async def retrieve_text_embeddings(
+        self, query: str, n: int = 10
+    ) -> List[Tuple[int, float]]:
+        """Retrieve embeddings related to the given query.
+
+        Parameters:
+        - query: The query to retrieve matching embeddings for.
+        - n: The number of embeddings to retrieve.
+
+        Returns:
+        List of `(chunk_id, score)` pairs from the embeddings.
+        """
+        embedded_query = await self._embedding.aget_text_embedding(query)
+
+        async with self._pg_pool.acquire() as conn:
+            logger.info("Executing SQL query for chunks from {}", self.collection_id)
+            embeddings = await conn.fetch(self._retrieve_embeddings, embedded_query, n)
+            embeddings = [e["chunk_id"] for e in embeddings]
+            return embeddings
+
+    async def retrieve_text_chunks(self, query: str, n: int = 10) -> List[TextChunk]:
+        """Retrieve embeddings related to the given query.
+
+        Parameters:
+        - query: The query to retrieve matching embeddings for.
+        - n: The number of embeddings to retrieve.
+
+        Returns:
+        List of chunk_ids from the embeddings.
+        """
+        embedded_query = await self._embedding.aget_text_embedding(query)
+
+        async with self._pg_pool.acquire() as conn:
+            logger.info("Executing SQL query for chunks from {}", self.collection_id)
+            embeddings = await conn.fetch(self._retrieve_chunks, embedded_query, n)
+            embeddings = [
+                TextChunk(raw=True, score=e["score"], text=e["text"])
+                for e in embeddings
+            ]
+            return embeddings
+
+    async def ingest(self, document_id: int, url: str) -> None:
+        logger.info("Loading content for document {} from '{}'", document_id, url)
+        extracted = await extract(
+            url, extract_tables=self.extract_tables, extract_images=self.extract_images
+        )
+        if extracted.is_empty():
+            logger.error(
```

Comment:
If this is an error, shouldn't it throw an exception?

Reply:
It could -- but with background tasks, there isn't really anything to *do* with that error. What I think we actually need to do is mark the document (or ingestion associated with the document) as failed and/or do some kind of dead letter. That said -- perhaps we shouldn't treat this as an error?